ENH Introduces set_output API for pandas output

thomasjpfan opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Closes https://github.com/scikit-learn/scikit-learn/issues/23001
Implements SLEP018: https://github.com/scikit-learn/enhancement_proposals/pull/68

#### What does this implement/fix? Explain your changes.
This PR introduces:

1. `set_output` for `Pipeline` and transformers in `preprocessing`.
2. Common test to check the behavior of `set_output`, which will be used for follow up PRs that adds `set_output` to all other transformers.
3. `OutputTypeMixin`, where most transformers only need to subclass it to get the `set_output` API.
4. Global configuration option `set_config(output_transform="pandas")` to set the output globally.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thomasjpfan committed:
ENH Introduces set_output API

thomasjpfan committed:
CLN Reduces API surface

thomasjpfan committed:
ENH Expand test for failing case

thomasjpfan committed:
CLN Rename mixin

thomasjpfan committed:
ENH Add full support for get_output in preprocessing

thomasjpfan committed:
DOC Adds comment in clone

thomasjpfan committed:
DOC Adds whats new number

thomasjpfan committed:
CLN Less diff

thomasjpfan committed:
ENH Use keyword only arguments for public API

ogrisel commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,225 @@
+
+    def set_output(self, *, transform=None):
+        """Set output container.
+
```
I think we should mark this method experimental to let the user known that the API might change without going through the usual deprecation cycle at this point.

thomasjpfan replied:
This method been marked as experimental with a `..note::`

thomasjpfan commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,225 @@
+            output.columns = columns
+        if index is not None:
+            output.index = index
+        return output
```
> should it be the responsibility of the implementer of that transformer to ensure consistency with get_feature_names_out

I'll prefer `get_feature_names_out` to already be consistent when it gets to wrapping the container. In that case, we can have `SetOutputMixin` be a no-op if the return of `transform` is already a DataFrame.

> Or is it just to add some compat with third-party estimators that output dataframes by default in case they plan to inherit from the SetOutputMixin?

This `if-block` is for compat, but it's also to make sense of the API. If the `output` is already a dataframe and `columns` is passed in, then I thought a reasonable thing to do is to set the `columns` as well. Given the point, above I am okay with doing a no-op if the container is already a DataFrame.

thomasjpfan replied:
I updated the PR to make it a noop when the `original_data` is a pandas dataframe. This way `SetOutputMixin` does not do anything if `estimator.transform` already returns a DataFrame.

lorentzenchr replied:
Is this resolved?

thomasjpfan replied:
I think there is still a lingering issue with `FunctionTransformer`.

This PR has been updated so that `TranformerMixin` inherits the `SetOutputMixin`. This means `FunctionTransformer` will get the `set_output` API if `feature_names_out` is defined. To be concrete, here is the current behavior for `FunctionTransformer`:

1. If `feature_names_out` is defined, then `set_output` is available. The output columns will always be consistent with `get_feature_names_out` no matter if `func` returns a dataframe or ndarray.

2. If `feature_names_out` is not defined then `set_output` is **not** available. `func` is allow to return a dataframe or a ndarray. Not having `set_output` is inconvenient, specifically when `func` returns a dataframe. For example, if the `FunctionTransformer` without `set_output` is in complex pipeline, then `pipeline.set_output(transform="pandas")` would fail to configure `FunctionTransformer`.

A workaround for 2 is to have a custom `set_output` raise a warning when `feature_names_out` is not defined. The warning states that `func` must return a dataframe if `set_output(transform="pandas")`. Technically, we can "learn" the feature names out by running `func` in `fit`, but that introduces more computation compared to `main`.

ogrisel replied:
> A workaround for 2 is to have a custom set_output raise a warning when feature_names_out is not defined. The warning states that func must return a dataframe if set_output(transform="pandas").

Sounds reasonable. We can also raise an error at transform time if `set_ouput(transform="pandas")` was previously called and that `func` does not naturally return a pandas dataframe.

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
FIX Fixes typo

thomasjpfan committed:
CLN Use dictionary instead

thomasjpfan committed:
CLN Better error message

thomasjpfan committed:
TST Adds more code coverage

thomasjpfan committed:
CLN Simplifies implementation

thomasjpfan committed:
CLN Remove unneeded parameter

thomasjpfan committed:
CLN Simplify validation

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

amueller commented on doc/whats_new/v1.2.rst:
```diff
@@ -39,6 +39,9 @@ random sampling procedures.
 -----------------------------
 
+- |MajorFeature| The `set_output` API has been adopted by transformers
+  in :mod:`sklearn.preprocessing`. :pr:`23734` by `Thomas Fan`_.
```
and Pipeline? Not impacting all modules strictly speaking but kind of?

amueller commented on sklearn/_config.py:
```diff
@@ -13,6 +13,7 @@
         os.environ.get("SKLEARN_PAIRWISE_DIST_CHUNK_SIZE", 256)
     ),
     "enable_cython_pairwise_dist": True,
+    "output_transform": "default",
```
it's ``transform_output`` in the SLEP right? That seems a bit nicer?

amueller commented on sklearn/preprocessing/_discretization.py:
```diff
@@ -11,15 +11,19 @@
 from . import OneHotEncoder
 
 from ..base import BaseEstimator, TransformerMixin
+<<<<<<< HEAD
```
merge error

amueller reviewed (commented):
I think we should move the `SetOutputMixin` into the `TransformerMixin` so we don't have to add it everywhere manually and to keep the inheritance list more reasonable.
That should work in theory, though things that output sparse will error on ``transform`` if someone did ``set_output`` on them. It might be nicer to error earlier, but we can still do that after merging the initial thing.

thomasjpfan committed:
CLN Rename output_transform transform_output

thomasjpfan committed:
CLN Fix name

thomasjpfan committed:
DOC Update whats new

amueller commented:
Ok so the problem with inheriting from ``TransformerMixin`` is that it's hard to overwrite the behavior done in the mixin.
With the implementation in this PR, transformers can decide not to inherit from `SetOutputMixin` and implement ``transform`` and ``fit_transform`` themselves to support ``set_output``.

If we add ``SetOutputMixin`` to ``TransformerMixin`` everybody inheriting from ``TransformerMixin`` will automatically get ``SetOutputMixin`` but it's harder to customize.
There would be two ways to allow customization, either via making  ``_wrap_method_output`` a method that someone could overwrite, or by adding a ``**kwargs`` to ``__init_subclass__`` which would allow someone to disable the automatic wrapping behavior.

Let the bikeshedding begin!

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
ENH More flexible pandas output

thomasjpfan committed:
ENH Fixes set_output in other transformers

thomasjpfan committed:
TST Adds failing test for column transformer

thomasjpfan committed:
ENH Give column transformer desired behavior

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
TST Fixes future warning

thomasjpfan committed:
CLN Refactor wrapper to take original input

thomasjpfan committed:
CLN Clean up column transformer

thomasjpfan committed:
CLN Improve function and parameter names

thomasjpfan committed:
ENH Adds support for cross decomposition

thomasjpfan committed:
CLN Smaller diff by removing unneeded feature

thomasjpfan committed:
TST Adds tests for no get_feature_names_out

thomasjpfan committed:
CLN Make get_output_config more flexible

thomasjpfan commented:
This PR was updated:

1. `TransformerMixin` now inherits from `SetOutputMixin`, which means many transformers will get `set_output` automatically, which is both a pro and a con. I am open to reverting this inheritance.
2. `set_output` is only available only when `get_feature_names_out` is defined.
3. `SetOutputMixin` includes an `auto_wrap_output` to disable the auto wrapping behavior.
4. `ColumnTransformer` now outputs DataFrames with mixed types. I had to pull in the fix from #23993 for `FunctionTransformer` to get `passthrough` to always return a DataFrame.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
DOC Update whats new

thomasjpfan committed:
CLN Reduce diff

glemaitre commented on sklearn/compose/_column_transformer.py:
```diff
@@ -253,6 +255,32 @@ def _transformers(self, value):
             self.transformers = value
 
+    def set_output(self, transform=None):
+        """Set output container.
```
It is a bit unsual that we are acting (and here mutate) on the estimators passed in the constructor. We usually do not do that.

I assume that the behaviour here would be the one expected: no one would like to specify the output for each estimator individually. I assume that we need to be the most explicit possible.

thomasjpfan replied:
I updated the docstring.

This behavior was explicitly stated in [SLEP018](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep018/proposal.html#detailed-description):

> Meta-estimators that support set_output are required to configure all inner transformers by calling set_output. Specifically all fitted and non-fitted inner transformers must be configured with set_output

As you stated, this feature is for the improved ergonomics.

glemaitre commented on sklearn/compose/_column_transformer.py:
```diff
@@ -803,6 +847,10 @@ def _hstack(self, Xs):
             return sparse.hstack(converted_Xs).tocsr()
         else:
             Xs = [f.toarray() if sparse.issparse(f) else f for f in Xs]
+            config = get_output_config("transform", self)
```
`fit_transform` will also called this function and we decide to ouput only depending on `transform`. It is a bit inconsistent. I don't know if we should change the code or just revise the SLEP to have one entry in the config for `fit_transform` and `transform`. Basically, we will never have anything for `fit` since it returns only `self`.

thomasjpfan replied:
I think this is expected behavior. If someone sets the output of `transform`, then I think it's okay to assume they want `fit_transform` to also have the same output.

For me, having to set both keys is a worst user experience.

glemaitre commented on sklearn/compose/_column_transformer.py:
```diff
@@ -543,14 +583,18 @@ def _update_fitted_transformers(self, transformers):
                 trans = "drop"
             elif old == "passthrough":
                 # FunctionTransformer is present in list of transformers,
                 # so get next transformer, but save original string
```
I think this line refer to `trans="passthrough"`. If both could be next, it would be better.

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -147,6 +149,23 @@ def __init__(self, steps, *, memory=None, verbose=False):
         self.verbose = verbose
 
+    def set_output(self, transform=None):
+        """Set output container.
```
I think that we should also mentioned that we mutate the original estimators. Here it is less surprising because the pipeline already mutate `steps` inplace :)

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -969,6 +988,24 @@ def __init__(
         self.verbose = verbose
 
+    def set_output(self, transform=None):
+        """Set output container.
```
same here.

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -4133,11 +4133,79 @@ def check_param_validation(name, estimator_orig):
                     continue
 
                 with raises(ValueError, match=match, err_msg=err_msg):
-                    if any(
```
Are you sure of removing this part?

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -4133,11 +4133,79 @@ def check_param_validation(name, estimator_orig):
+    if name in CROSS_DECOMPOSITION:
+        X_trans_default = X_trans_default[0]
+
+    # Default and no setting returne the same transform
```
```suggestion
    # Default and no setting -> returns the same transform
```

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -4133,11 +4133,79 @@ def check_param_validation(name, estimator_orig):
+        X_trans_pandas = transformer.fit_transform(df, y)
+    except ValueError as e:
+        # transformer does not support sparse data
+        assert str(e) == "Pandas output does not support sparse data", e
```
It could be worth to instantiate some estimator with `sparse_output=False` to have a better coverage.

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -4133,11 +4133,79 @@ def check_param_validation(name, estimator_orig):
+    if name in CROSS_DECOMPOSITION:
+        X_trans_pandas = X_trans_pandas[0]
+
+    # Return is a pandas DataFrame
```
```suggestion
    # X_trans_pandas is a pandas DataFrame
```

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -4133,11 +4133,79 @@ def check_param_validation(name, estimator_orig):
+        X_trans_pandas = X_trans_pandas[0]
+
+    # Return is a pandas DataFrame
+    assert hasattr(X_trans_pandas, "iloc")
```
since the test is specialized for pandas and we already imported pandas, we could check that this is definetely a pandas dataframe.

glemaitre replied:
When I see the assertions below, I think that we could even use https://pandas.pydata.org/docs/reference/api/pandas.testing.assert_frame_equal.html to check all of those in one line.

glemaitre reviewed (commented):
A couple of comments. Some are only nitpicking.

glemaitre commented:
I did not look yet at some of the tests (apart from the common tests). I will have a go on them. I think that we miss documentation (for our users and developer).

Basically, the points that could raise some discussions are:

- we are mutating the passed estimators in the pipeline and column transformer. Are we fine with this?
- for the config, we have a flat dict. It could make it easier to work with nested dict?
- in the config we store 2 keys: one `transform` and one for `fit_transform`. I would merge both stating that 1 key for `*transform` is more than enough. `fit` will always return `self`. The same would apply for `fit_predict` and `predict`. I don't see the use case that one wants a dataframe at `transform` but not at `fit_transform`.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
WIP Move avaliable_if it its own file

thomasjpfan committed:
CLN Be more strict about columns existing

thomasjpfan commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,290 @@
+    def __init_subclass__(cls, auto_wrap_output=True, **kwargs):
+        # Dynamically wraps `transform` and `fit_transform` and configure it's
+        # output based on `set_output`.
+        if not isinstance(auto_wrap_output, bool):
```
There was support for callable, but found that it was not needed. In the current state, `auto_wrap_output` only accept bools.

lorentzenchr commented on doc/whats_new/v1.2.rst:
```diff
@@ -41,6 +41,9 @@ random sampling procedures.
 Changes impacting all modules
 -----------------------------
 
+- |MajorFeature| The `set_output` API has been adopted by all transformers.
```
Maybe also link to the SLEP.

lorentzenchr commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,290 @@
+
+    Parameters
+    ----------
+    data_to_wrap : ndarray, sparse matrix or pandas DataFrame
```
Let's start with raising an error. The tedious manual setting of several `sparse_output` options can be addressed in later PRs. Not raising errors anymore is not an API break IIUC :smirk:

thomasjpfan replied:
This PR has went through a few iterations and I missed this docstring. This PR raises an error for sparse matrices, which is the behavior explicitly stated in [SLEP018](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep018/proposal.html#detailed-description).

lorentzenchr commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,290 @@
+    )
+
+
+class SetOutputMixin:
```
Should here also state
```
.. note:: Experimental API
```
???

lorentzenchr commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,225 @@
+]
+
+
+def make_named_container(
```
Everything in `set:output.py` is private (leading underscore), except for `safe_set_output`, `SetOutputMixin` and `get_output_config`.

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
CLN Address more comments

thomasjpfan committed:
REV Revert unneeded change

thomasjpfan committed:
TST Ust assert_from_equal

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
TST Adds OneHotEncoder(sparse_output=True) to common test

thomasjpfan committed:
CLN Better column transformer logic

thomasjpfan commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,290 @@
+        Estimator instance.
+    """
+    set_output_for_transform = (
+        hasattr(estimator, "transform")
+        or hasattr(estimator, "fit_transform")
+        and transform is not None
```
I think this type of refactor can come after we decide on how to extend the API for other methods.

This functions refactor depends on how we want to extend `set_output` in the future. For example, there are two API options for `LogisticRegression.set_output` to configure `predict`:

1. `set_output` accepts `transform`, but it noops for it. This is the easiest. In this case, we can pass `transform=None` to `LogisticRegression.set_output` and it does nothing with ti.
2. `set_output` does not accept `transform`. In this case, `safe_set_output` needs to be smarter about passing kwargs around.

thomasjpfan committed:
CLN More strict language in error message

thomasjpfan committed:
DOC Remove docstring for unused parameter

thomasjpfan committed:
TST More tests

thomasjpfan committed:
DOC Better docstring

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
TST Fix failing tests

thomasjpfan committed:
DOC Adds example on using set_output API

thomasjpfan committed:
DOC Add developer docs for set_output

thomasjpfan committed:
DOC Adds more developer docs

thomasjpfan committed:
FIX Fixes bug with set_output and transform=None

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
ENH Moves set_output mixin into BaseEstimator

thomasjpfan committed:
API Make more API private

thomasjpfan committed:
DOC Remove unneeded link

thomasjpfan commented:
This PR was updated with:

1. Moved `avaliable_if` to it's own file to prevent circular dependency
2. Always require columns: https://github.com/scikit-learn/scikit-learn/pull/23734#discussion_r968386171
3. `SetOutputMixin` is now private. Classes that inherit from `BaseEstimator`  will get the feature right away.
4. Add documentation for users and developers
5. Moved all the utility functions as private.

Things I need to think about:
1. FunctionTransformer that already returns a DataFrame: https://github.com/scikit-learn/scikit-learn/pull/23734#discussion_r978181639
2. Using `sklearn.experimental` to be a feature flag. I can get this to work for `set_output` which is a user facing API. I'm deciding if we need to mark `set_output` as experimental at all. Currently, the API is fairly simple. The only change I see to it is adding `predict="pandas"` or enabling sparse support. Neither of these are backward breaking.

thomasjpfan committed:
API Move back to TransformerMixin

thomasjpfan committed:
ENH FunctionTransformer.set_output warns about func when transform=pandas

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

lorentzenchr commented on sklearn/feature_extraction/text.py:
```diff
@@ -567,7 +567,9 @@ def _warn_for_unused_params(self):
 
-class HashingVectorizer(TransformerMixin, _VectorizerMixin, BaseEstimator):
+class HashingVectorizer(
+    TransformerMixin, _VectorizerMixin, BaseEstimator, auto_wrap_output=False
```
```suggestion
    TransformerMixin, _VectorizerMixin, BaseEstimator, auto_wrap_output_keys=None
```
Typo or did I miss something?

lorentzenchr commented on sklearn/preprocessing/_function_transformer.py:
```diff
@@ -306,3 +306,35 @@ def __sklearn_is_fitted__(self):
+
+        if transform == "pandas" and self.feature_names_out is None:
+            warnings.warn(
+                'With transform="pandas" `func` should return a DataFrame to follow'
```
```suggestion
                'With transform="pandas", `func` should return a DataFrame to follow'
```

lorentzenchr commented on sklearn/utils/estimator_checks.py:
```diff
@@ -4145,3 +4145,77 @@ def check_param_validation(name, estimator_orig):
+        X_trans_no_setting = X_trans_no_setting[0]
+
+    transformer.set_output(transform="default")
+    X_trans_default = transformer.fit_transform(X, y)
```
Should we also test sequential `fit().transform()`?

thomasjpfan replied:
That makes sense. I added testing for `fit().transform()` in [`1421e8a` (#23734)](https://github.com/scikit-learn/scikit-learn/pull/23734/commits/1421e8ae232bf6f68a53361486511faaaff8f7ba)

lorentzenchr commented on sklearn/compose/_column_transformer.py:
```diff
@@ -303,7 +334,14 @@ def _iter(self, fitted=False, replace_strings=False, column_as_strings=False):
         """
         if fitted:
-            transformers = self.transformers_
+            # Replace "passthrough" with the fitted version in
+            # _name_to_fitted_passthrough
+            def replace_passthrough(name, trans, columns):
+                if not replace_strings or name not in self._name_to_fitted_passthrough:
+                    return name, trans, columns
+                return name, self._name_to_fitted_passthrough[name], columns
+
+            transformers = [replace_passthrough(*trans) for trans in self.transformers_]
```
```suggestion
            if replace_strings:
                # Replace "passthrough" with the fitted version in
                # _name_to_fitted_passthrough
                def replace_passthrough(name, trans, columns):
                    if name not in self._name_to_fitted_passthrough:
                        return name, trans, columns
                    return name, self._name_to_fitted_passthrough[name], columns

                transformers = [replace_passthrough(*trans) for trans in self.transformers_]
            else:
                transformers = self.transformers_
```
Just a suggestion to shortcut for `replace_strings=False`.

lorentzenchr reviewed (commented):
This is looking very good. Apart from minor review comments, I've the following questions:
1. How difficult/easy would it be, to extend the `set_output` options to, e.g., `"arrow"`.
2. What happens with 3rd party transformers in a pipeline if global `set_config(transform_output="pandas")` has been set?
3. If `set_config(transform_output="pandas")` has been set, and a model is saved via pickle and later loaded in a different python process, will it ouput pandas?
4. In the default setting, what happens with other Array API compliant data containers than numpy, e.g. cupy, see https://github.com/scikit-learn/scikit-learn/issues/22352?

thomasjpfan commented on sklearn/_config.py:
```diff
@@ -14,6 +14,7 @@
     ),
     "enable_cython_pairwise_dist": True,
     "array_api_dispatch": False,
+    "transform_output": "default",
```
This option use to be `"numpy"`, but it was inconsistent with transformers that can output sparse data. I was also thinking about third party transformers that already output dataframes where a "numpy" default would be strange.

I can get behind `"array-like"`. The only concern I have is how sparse data is a weird "array-like", because `asarray` on a sparse matrix returns an object dtype:

```python
from scipy.sparse import csr_matrix
import numpy as np

mat = csr_matrix([[1, 2, 0]])
print(np.asarray(mat).dtype)
# object
```

thomasjpfan replied:
Looking at our [glossary entry](https://scikit-learn.org/stable/glossary.html#term-array-like) for "array-like", we exclude "sparse matrix" from "array-like". In that case, "array-like" would not be a good default, because it does not cover sparse matrices.

lorentzenchr replied:
Do you have a proposal. Naming the default "default" just seems wrong to me. What if we change it in the future?

thomasjpfan replied:
The semantics for "default" is "the transformer does anything it wants". Here are some options:

1. `None`
2. `"undefined"`
3. `"unmodified"`
4. `"unchanged"`

I am in favor of `None`. I think it's the most pythonic way to say "use the default".

lorentzenchr replied:
I'm fine with `None`. What do others think?

thomasjpfan committed:
Apply suggestions from code review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
CLN Smaller diff compared to main

thomasjpfan committed:
STY Flake8 error

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
FIX Merge conflict

thomasjpfan committed:
CLN Address more comments

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan committed:
FIX Fixes doc build

thomasjpfan committed:
FIX Fixes test for new message

thomasjpfan committed:
CLN Address comments

thomasjpfan committed:
DOC Adds comments about data_to_wrap

thomasjpfan committed:
DOC Adds behavior with multiple array output

thomasjpfan committed:
STY Slight formatting

thomasjpfan committed:
TST Adds testing for fit_transform and fit.transform

thomasjpfan commented on sklearn/utils/set_output.py:
```diff
@@ -0,0 +1,266 @@
+    def wrapped(self, X, *args, **kwargs):
+        data_to_wrap = f(self, X, *args, **kwargs)
+        if isinstance(data_to_wrap, tuple):
+            # only wrap the first output for cross decomposition
```
I added the behavior into the developer docs: [`d25ba8b` (#23734)](https://github.com/scikit-learn/scikit-learn/pull/23734/commits/d25ba8b7f07c2644204fd9635aeabca4f6f89f90)

lorentzenchr commented on doc/developers/develop.rst:
```diff
@@ -635,6 +635,35 @@ instantiated with an instance of ``LogisticRegression`` (or
+        def get_feature_names_out(self, input_features=None):
+            ...
+
+For transformers that return multiple array in `transform`, auto wrapping will
```
```suggestion
For transformers that return multiple arrays in `transform`, auto wrapping will
```

thomasjpfan commented:
> How difficult/easy would it be, to extend the set_output options to, e.g., "arrow".

I see two possible paths: Support it directly in scikit-learn or an API to configure "arrow".

1. Add another function `_wrap_in_arrow_container` and dispatch to it in `_wrap_data_with_container` when `set_output(transform="arrow")`.
2. Allow `set_output(transform=callable)`, where a users defines a callable that constructs an arrow container. This will likely have the same signature of `_wrap_in_pandas_container`: `(data, columns, index)`.

> What happens with 3rd party transformers in a pipeline if global `set_config(transform_output="pandas")` has been set?

If they inherit from `TransformerMixin`, then they will auto opt-into `set_output` and `transform` will output DataFrames. They can opt-out by setting `auto_wrap_output_keys=None`.

> If `set_config(transform_output="pandas")` has been set, and a model is saved via pickle and later loaded in a different python process, will it ouput pandas?

If it's a global `set_config`, then the global option needs to be set again to output pandas. If it's the local option, i.e. `transform.set_output`, then `transform` outputs pandas.

> In the default setting, what happens with other Array API compliant data containers than numpy, e.g. cupy, see [#22352](https://github.com/scikit-learn/scikit-learn/issues/22352)?

The default setting will keep the behavior on `main`. The transformer decides what it wants to output, ndarray, cupy, dataframes, etc.

lorentzenchr commented:
> >What happens with 3rd party transformers in a pipeline if global set_config(transform_output="pandas") has been set?

>If they inherit from TransformerMixin, then they will auto opt-into set_output and transform will output DataFrames. They can opt-out by setting auto_wrap_output_keys=None.

What if they do not inherit from `TransformerMixin`? Or stated otherwise: Are we clear enough on the expected API for a 3rd party transformer (that - for some reasons - does not want to depend on scikit-learn)?

lorentzenchr commented:
@thomasjpfan The only open point for me is the name of the default option of `transform_output`, see https://github.com/scikit-learn/scikit-learn/pull/23734#discussion_r985071192. Your suggestion to use `None` instead of `"default"` is good for me. As this would be a deviation from the SLEP, other opinions are highly appreciated.

Once that is settled, I will give my review approval.

thomasjpfan commented:
> What if they do not inherit from TransformerMixin? Or stated otherwise: Are we clear enough on the expected API for a 3rd party transformer (that - for some reasons - does not want to depend on scikit-learn)?

For now, I prefer not to require the API from third party estimators. Our meta-estimators such as `ColumnTransformer` will complain when transformers `set_output`. As for the global option, I do not think we can require third parties to respect it.

Adopting the `set_output` API fully, requires at least a soft dependency on `scikit-learn` and `pandas`:

- `scikit-learn`: To follow scikit-learn's global config, one needs to get the global config from scikit-learn. 
- `pandas`: Outputting DataFrames requires it.

thomasjpfan committed:
Update doc/developers/develop.rst

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

glemaitre reviewed (approved):
I am a bit split regarding the choice between `None`, `"default"` (why not `"native"`). I find that `None` does not specify the output type but this is also the case for `"default"` or `"native"`. It means that the output type should be provided somewhere else (I assume the documentation).

I am therefore fine with any option.

Regarding the other changes. LGTM.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pandas_out_prototype_v3

thomasjpfan commented:
Looking through the code, we can not use `None` instead of `"default"`. Currently, the signature is: `set_output(self, transform=None)`, where `None` is a sentinel to mean do "not configure transform". This is required if we do `set_output(predict="pandas")`, which configures the container for `predict` but leaves `transform` alone. Secondly, `est.set_output()` with no input also configures nothing.

I can use another sentinel object, but that complicates the `set_output` API.

I think the best option is to find a string that best describes the behavior. I think "native" is okay. Another options is: "any"

@lorentzenchr WDYT?

lorentzenchr commented:
Yes `"native"` is fine, even good! I'm happy with any other default value than `"default"`, which I would consider bad design because it does not tell anything about the actual behaviour.

thomasjpfan committed:
ENH Rename 'default' to 'native'

thomasjpfan commented:
I updated this PR to use `"native"`. I also opened a PR to update SLEP: https://github.com/scikit-learn/enhancement_proposals/pull/78

lorentzenchr commented:
> I updated this PR to use `"native"`.

Did you forget to push?

lorentzenchr reviewed (approved):
LGTM
@thomasjpfan A great thank you for all your (years long) efforts to make this happen!

thomasjpfan committed:
Revert "ENH Rename 'default' to 'native'"

This reverts commit 531395838327a636042bc8bd93d2a5c54c6c0007.

lorentzenchr commented:
I merge. In case that [#78](https://github.com/scikit-learn/enhancement_proposals/pull/78) concludes to still change the default value, we can do that in a new PR.

amueller commented:
Yaaaaay!!
